pubsub/pubsub.go

```diff
@@ -0,0 +1,211 @@
+	ackID AckID
+
+	// sub is the Subscription this message was received from.
+	sub *Subscription
+}
+
+type AckID interface{}
+
+// Ack acknowledges the message, telling the server that it does not need to
+// be sent again to the associated Subscription. This method blocks until
+// the message has been confirmed as acknowledged on the server, or failure
+// occurs.
+func (m *Message) Ack(ctx context.Context) error {
+	// Send the ack ID back to the subscriber for batching.
+	m.sub.ackChan <- m.ackID
+	select {
+	case err := <-m.sub.ackErrChan:
+		return err
+	case <-ctx.Done():
+		return nil
+	}
+}
+
+// Topic publishes messages to all its subscribers.
+type Topic struct {
+	driver   driver.Topic
+	mcChan   chan msgCtx
+	doneChan chan struct{}
+}
+
+// TopicOptions contains configuration for Topics.
+type TopicOptions struct {
+	// SendDelay tells the max duration to wait before sending the next batch of
+	// messages to the server.
+	SendDelay time.Duration
+
+	// BatchSize specifies the maximum number of messages that can go in a batch
+	// for sending.
+	BatchSize int
+}
+
+// msgCtx pairs a Message with the Context of its Send call.
+type msgCtx struct {
+	msg *Message
+	ctx context.Context
+}
+
+// Send publishes a message. It only returns after the message has been
+// sent, or failed to be sent. Send can be called from multiple goroutines
+// at once.
+func (t *Topic) Send(ctx context.Context, m *Message) error {
+	t.mcChan <- msgCtx{m, ctx}
+	// Wait for the batch including this message to be sent to the server.
+	return <-m.errChan
+}
+
+// Close disconnects the Topic.
+func (t *Topic) Close() error {
+	close(t.doneChan)
+	return t.driver.Close()
+}
+
+// NewTopic makes a pubsub.Topic from a driver.Topic and opts to
+// tune how messages are sent. Behind the scenes, NewTopic spins up a goroutine
+// to bundle messages into batches and send them to the server.
+func NewTopic(d driver.Topic, opts TopicOptions) *Topic {
+	t := &Topic{
+		driver:   d,
+		mcChan:   make(chan msgCtx),
+		doneChan: make(chan struct{}),
+	}
+	go func() {
+		// Pull messages from t.mcChan and put them in batches. Send the current
+		// batch whenever it is large enough or enough time has elapsed since
+		// the last send.
+		for {
+			batch := make([]*driver.Message, 0, batchSize)
+			timeout := time.After(opts.SendDelay)
+		Loop:
+			for i := 0; i < opts.BatchSize; i++ {
+				select {
+				case <-timeout:
+					// Time to send the batch, even if it isn't full.
+					break Loop
+				case mc := <-t.mcChan:
+					select {
+					case <-mc.ctx.Done():
+						// This message's Send call was cancelled, so just skip
+						// over it.
+					default:
+						dm := &driver.Message{
+							Body:       m.msg.Body,
+							Attributes: m.msg.Attributes,
+							AckID:      m.msg.AckID,
+						}
+						batch = append(batch, dm)
+					}
+				case <-t.doneChan:
+					return
+				}
+			}
+			if len(batch) > 0 {
+				err := t.driver.SendBatch(ctx, batch)
```

Comment:
It will have to come from a parameter of NewTopic. It can't be the ctx of any particular message because then the cancellation of that one would cancel sending the whole batch.